labs/KeyWrapping/KeyWrappingAnswers.md

```diff
@@ -0,0 +1,1168 @@
+shrink a = drop a
+```
+
+`widen` takes any sized input and prepends `0` or more `False`
+bits. `shrink` takes any sized input and removes `0` or more bits from
+the front. Using these two functions, we can fix our `f` from above:
+
+```
+f : {a} (32 <= a, a <= 64) => [a] -> [48]
+f x = if `a <= 0x30 then
+        widen (g (shrink x))
+      else
+        shrink (h (widen x))
+```
+
+And here we test that `f` correctly calls `g` and `h` (which increment
+and decrement by 1, respectively).
+
+```sh
+labs::KeyWrapping::KeyWrapping> f (10 : [37])
+11
+labs::KeyWrapping::KeyWrapping> f (10 : [53])
+9
+```
+
+
+## KWP-AE Top Level Function
+
+With those two considerations firmly under our belt, we can now tackle
+`KWP-AE`.
+
+**EXERCISE**: Study `Algorithm 5` from the standard and complete the
+  definition of `KWPAE` below by filling in the function skeleton
+  provided with appropriate logic. Use the `KWPAEPad` function from
+  above to create `S`. Use the `shrink` and `widen` functions to
+  assist in resizing `S` and the function outputs on the `then` and
+  `else` branches of line 5.
+  
+*Hint:* You'll notice that we needed to pull in the type variable `n`
+and type constraints from `W` and relate `n` and `l` (the type of both
+`S` and `C`). It may also be necessary to tell `W` that our type
+variable `n` is the same type variable that it uses. This can be
+achieved by calling `W` like so: ```W`{n=n}```.
+
+```
+KWPAE :
+    {k, l, n}            // k is [len(P)/8], Algorithm 5
+    ( 1 <= k, k < 2^^32  // Bounds on the number of octets of P, from Table 1
+    , l == 32 + 32 + k*8 + k*8 %^ 64  // The type of S and C
+    , 64*n == max 192 l               // Here we relate n and l
+    ) =>
+    ([128] -> [128]) -> [k][8] -> [l]
+KWPAE CIPHk P = C
+  where
+    S = KWPAEPad P
+    C = if (`k : [32]) <= 8 then
+          widen (CIPHk (shrink S))
+        else
+          shrink (join (W`{n} CIPHk (split (widen S))))
+```
+
+Feel free to use the provided `KWPAETests` property to check your
+work.
+
+
+# A Formal Specification of `KWP-AD`
+
+`KWP-AD` is the authenticated-decryption function and makes use of our
+previously defined `W'`.
+
+**EXERCISE**: Study `Algorithm 6` from the standard and complete the
+  definition of `KWPAD` below by filling in the function skeleton
+  provided with appropriate logic. We suggest splitting the algorithm
+  into two again, so we're providing a skeleton for `KWPADUnpad` (which
+  is roughly the inverse of `KWPAEPad`) and `KWPAD`.
+
+```
+KWPADUnpad :
+    {k, l}               // k is [len(P)/8], Algorithm 5
+    ( 1 <= k, k < 2^^32  // Bounds on the number of octets of P, from Table 1
+    , l == 32 + 32 + k*8 + k*8 %^ 64  // The type of S
+    ) =>
+    [l] -> (Bit, [k][8])
+KWPADUnpad S = (FAIL, split P)
+  where
+    MSB32 : [32]
+    k     : [32]
+    P     : [k*8]
+    PAD   : [k*8 %^ 64]
+    MSB32 # k # P # PAD = S
+    FAIL = MSB32 != ICV2 \/
+           k     != `k   \/
+           PAD   != 0
+```
+
```

Comment:
Does this actually help us do step 7 (If padlen < 0 or padlen > 7, then return FAIL and stop)? Are we just assuming that the function wouldn't work if, for example, the padding added to the message was more than `[k*8 %^ 64]`  bits? Maybe that's really particular and we don't actually care.

Reply:
Also, have we mentioned in this lab or another that 'k refers to the value of the type variable k? so the comparison of 'k will make sure we got the same # of bytes from the output.
Ohhh this check probably actually helps if we had super long padding.

Reply:
yeah!